### ISSUE TITLE
Handling file/buffer rename

### ISSUE THREAD

**Issue by ruricolist** [joaotavora/eglot#522]
This isn't a bug, but something that took me by surprise: if I open a file in Eglot and then rename the file/buffer (with Dired,) Eglot does not send textDocument/didClose for the old file name or textDocument/didOpen events for the new file name.

I would expect that handling this would require advice on set-visited-file-name.

**Comment by joaotavora** [comment]
Kind of sounds a bug, though :-). Doesn't it confuse the server? I.e. can you continue working afterwards?

**Comment by ruricolist** [comment]
Yes, unfortunately it confuses the server and it has to be restarted.

**Comment by skangas** [comment]
I can reproduce this using pyls, steps to reproduce:
1. C-x C-f /tmp/foo.el
2. M-x eglot
3. C-x C-j and rename the file in dired
4. Go back to the now renamed buffer and see that it doesn't work.

**Comment by joaotavora** [comment]
Yeah, for this be fixed, Eglot has to hook on whatever mechanism Emacs uses to rename files.

However, I do know that some servers also "watch" directories themselves , or rather they ask the client to watch the files/directories for them.  Eglot supports this.  IT's called the `workspace/didChangeWatchedFiles` thing. `pyls` doesn't support it, but `pyright` might.  So a "better" server might support this

**Comment by skangas** [comment]
I have a finished patch against Emacs master to add a new `before-set-visited-file-name-hook` that we could use for this in Emacs 29, while in Emacs 28 or earlier we'd need to do a bit more legwork ourselves (read: it'll be more ugly, but also behind `(if (boundp 'before-set-visited-file-name-hook) ...)`). I could finish up the patch for `eglot` very quickly too, I think. That would work for all renames in Emacs (from Dired, etc.).

But now that you mention doing this with LSP, actually that might not be such a bad idea. For example, if the file is renamed from the terminal while it's visited in Emacs, it would be cool if eglot could pick up on that and automatically rename your buffer for you... But that is more on the "nifty" side than the "useful" one perhaps, and maybe even surprising enough that it should be optional.

Maybe we should start with only caring about renames from Emacs itself, and add the `workspace/didChangeWatchedFiles` stuff later?

**Comment by joaotavora** [comment]
> Maybe we should start with only caring about renames from Emacs itself, and add the workspace/didChangeWatchedFiles stuff later?

yeah.  I'm not sure if you understood (or if I'm explaining correctly).  This workspace/didChangeWatchedFiles is the regular LSP mechanism whereby _some_ servers ask the client to install filesystem watchers and tell them which FS events happened in particular directories.  Why it is not the server that does this, it baffles me to an extent, but maybe it's because theoretically servers can run remotely to the files being edited (but then, so can clients 🤷 )

Anyway, that feature is _already_ supported by Eglot, as far as I can tell.  There's nothing to add.  So its possible that if you had tried that test with another, a "better" server, you simply wouldn't have seen this bug.

In that hypothetical, to-be-confirmed case, I would personally be confortable as writing this off as a "server bug" and _not_ add any complexity to eglot, which, as you note, is hard to do correctly (because renames can come from multiple places).

**Comment by skangas** [comment]
TL;DR. I will reply back once I had time to experiment more with this.

> Anyway, that feature is _already_ supported by Eglot, as far as I can tell. There's nothing to add. So its possible that if you had tried that test with another, a "better" server, you simply wouldn't have seen this bug.

Your previous explanation was clear, but I misread. I will try to find some server that supports this to see how this works.

> In that hypothetical, to-be-confirmed case, I would personally be confortable as writing this off as a "server bug" and _not_ add any complexity to eglot, which, as you note, is hard to do correctly (because renames can come from multiple places).

With the new hook I would add to Emacs 29, this would become almost trivial. I think. However, the question is if the built-in support and the `workspace/didChangeWatchedFiles` would start stepping on each others toes. If that is the case, I would suggest we should just write this up as a "server bug" and do nothing.

**Comment by joaotavora** [comment]
> With the new hook I would add to Emacs 29, this would become almost trivial. I think. However, the question is if the built-in support and the workspace/didChangeWatchedFiles would start stepping on each others toes.

Right: toe-stepping must be watched out for, makes perfect sense :+1: and good plan.
